libs/widget-lib/src/cowSwapWidget.ts

```diff
@@ -49,19 +54,36 @@ export function cowSwapWidget(
 
   applyDynamicHeight(iframe, params.height)
 
-  if (provider) {
-    const jsonRpcManager = new JsonRpcManager(contentWindow)
-
-    jsonRpcManager.onConnect(provider)
-  }
+  let jsonRpcManager = updateProvider(contentWindow, null, provider)
 
   iframe.addEventListener('load', () => {
-    updateWidget(params, contentWindow)
+    updateWidget(contentWindow, params)
   })
+  const iFrameCowEventEmitter = new IframeCowEventEmitter(listeners)
+
+  return {
+    updateWidget: (newParams: CowSwapWidgetParams) => updateWidget(contentWindow, newParams),
+    updateListeners: (newListeners?: CowEventListeners) => iFrameCowEventEmitter.updateListeners(newListeners),
+    updateProvider: (newProvider) => {
+      jsonRpcManager = updateProvider(contentWindow, jsonRpcManager, newProvider)
+    },
+  }
+}
+
+function updateProvider(contentWindow: Window, jsonRpcManager: JsonRpcManager | null, newProvider?: EthereumProvider) {
```

Comment:
I'm not a big fan of function arguments mutations, so I propose:

```
function updateProvider(contentWindow: Window, jsonRpcManager: JsonRpcManager | null, newProvider?: EthereumProvider) {
  const currentJsonRpcManager = jsonRpcManager || new JsonRpcManager(contentWindow)

  currentJsonRpcManager.disconnect()
  jsonRpcManager.onConnect(newProvider)
  
  return currentJsonRpcManager
}
```

Reply:
966b255a30730af310cf9e570b7bf4fb91e7c4b0